pkg/comp-functions/functions/vshn-postgres-func/mailgun_alerting.go

```diff
@@ -0,0 +1,133 @@
+	vshnv1 "github.com/vshn/appcat/apis/vshn/v1"
+	runtime "github.com/vshn/appcat/pkg/comp-functions/runtime"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/utils/pointer"
+	controllerruntime "sigs.k8s.io/controller-runtime"
+)
+
+func MailgunAlerting(ctx context.Context, iof *runtime.Runtime) runtime.Result {
+	log := controllerruntime.LoggerFrom(ctx)
+	log.Info("Starting mailgn-alerting function")
+
+	comp := &vshnv1.VSHNPostgreSQL{}
+	err := iof.Observed.GetComposite(ctx, comp)
+	if err != nil {
+		return runtime.NewFatalErr(ctx, "Cannot get composite from function io", err)
+	}
+	// Wait for the next reconciliation in case instance namespace is missing
+	if comp.Status.InstanceNamespace == "" {
+		return runtime.NewWarning(ctx, "Composite is missing instance namespace, skipping transformation")
+	}
+
+	if comp.Spec.Parameters.Monitoring.Email != "" {
+		log.Info("email not null")
+		err := deployAlertmanagerConfig(ctx, comp, iof)
+		if err != nil {
+			fmt.Println(err)
+			return runtime.NewFatalErr(ctx, "Can't add "+comp.Name+"-alertmanagerconfig-mailgun", err)
+		}
+	}
+
+	log.Info("Finishing mailgun-alerting function with NewNormal")
+
+	return runtime.NewNormal()
+}
+
+func deployAlertmanagerConfig(ctx context.Context, comp *vshnv1.VSHNPostgreSQL, iof *runtime.Runtime) error {
+	log := controllerruntime.LoggerFrom(ctx)
+	var alertManagerConfigName = comp.Name + "-alertmanagerconfig-mailgun"
+	var alertManagerConfigSecretName = comp.Name + "-alertmanagerconfig-mailgun-secret"
+	receiverName := "mailgun"
+
+	ac := &alertmanagerv1alpha1.AlertmanagerConfig{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      alertManagerConfigName,
+			Namespace: comp.Status.InstanceNamespace,
+			Labels: map[string]string{
+				"alertManagerConfig": "mailgun",
+			},
+		},
+		Spec: alertmanagerv1alpha1.AlertmanagerConfigSpec{
+			Receivers: []alertmanagerv1alpha1.Receiver{
+				{
+					Name: receiverName,
+					EmailConfigs: []alertmanagerv1alpha1.EmailConfig{
+						{
+							To:           comp.Spec.Parameters.Monitoring.Email,
+							From:         "appcat@appuio.cloud",
+							AuthUsername: "appcat@appuio.cloud",
+							AuthPassword: &v1.SecretKeySelector{
+								Key: "password",
+								LocalObjectReference: v1.LocalObjectReference{
+									Name: alertManagerConfigSecretName,
+								},
+							},
+							Smarthost:    "smtp.mailgun.org:587",
+							RequireTLS:   pointer.Bool(true),
+							SendResolved: pointer.Bool(true),
+						},
+					},
+				},
+			},
+			Route: &alertmanagerv1alpha1.Route{
+				GroupBy: []string{
+					"alertname",
+				},
+				GroupWait:      "10s",
+				GroupInterval:  "5m",
+				RepeatInterval: "1h",
+				Receiver:       receiverName,
+			},
+		},
+	}
+
+	secret := &v1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      alertManagerConfigSecretName,
+			Namespace: comp.Status.InstanceNamespace,
+		},
+		StringData: map[string]string{
+			"password": "takeitfromsecret",
```

Comment:
This is probably wrong? Do you need this line here? The reference should work without having a dumb value first.

Reply:
Removed it